src/couch_jobs/src/couch_jobs_fdb.erl

```diff
@@ -0,0 +1,676 @@
+    jiffy:decode(JobData, [return_maps]).
+
+
+% Cached job transaction object. This object wraps a transaction, caches the
+% directory lookup path, and the metadata version. The function can be used
+% from or outside the transaction. When used from a transaction it will verify
+% if the metadata was changed, and will refresh automatically.
+%
+get_jtx() ->
+    get_jtx(undefined).
+
+
+get_jtx(#{tx := Tx} = _TxDb) ->
+    get_jtx(Tx);
+
+get_jtx(undefined = _Tx) ->
+    case ets:lookup(?MODULE, ?JOBS_ETS_KEY) of
+        [{_, #{} = JTx}] ->
+            JTx;
+        [] ->
+            JTx = update_jtx_cache(init_jtx(undefined)),
+            JTx#{tx := undefined}
+    end;
+
+get_jtx({erlfdb_transaction, _} = Tx) ->
+    case ets:lookup(?MODULE, ?JOBS_ETS_KEY) of
+        [{_, #{} = JTx}] ->
+            ensure_current(JTx#{tx := Tx});
+        [] ->
+            update_jtx_cache(init_jtx(Tx))
+    end.
+
+
+% Transaction processing to be used with couch jobs' specific transaction
+% contexts
+%
+tx(#{jtx := true} = JTx, Fun) when is_function(Fun, 1) ->
+    fabric2_fdb:transactional(JTx, Fun).
+
+
+% Debug and testing API
+
+get_job(Type, JobId) ->
+    fabric2_fdb:transactional(fun(Tx) ->
+        JTx = init_jtx(Tx),
+        case get_job_val(Tx, job_key(JTx, Type, JobId)) of
+            #jv{seq = Seq, jlock = JLock} = JV ->
+                #{
+                    job => true,
+                    type => Type,
+                    id => JobId,
+                    seq => Seq,
+                    jlock => JLock,
+                    stime => JV#jv.stime,
+                    resubmit => JV#jv.resubmit,
+                    data => decode_data(JV#jv.data),
+                    state => job_state(JLock, Seq)
+                };
+            not_found ->
+                not_found
+        end
+    end).
+
+
+get_jobs() ->
+    fabric2_fdb:transactional(fun(Tx) ->
+        #{jobs_path := Jobs} = init_jtx(Tx),
+        Prefix = erlfdb_tuple:pack({?DATA}, Jobs),
+        Opts = [{streaming_mode, want_all}],
+        Result = erlfdb:wait(erlfdb:get_range_startswith(Tx, Prefix, Opts)),
+        lists:map(fun({K, V}) ->
+            {Type, JobId} = erlfdb_tuple:unpack(K, Prefix),
+            {Seq, JLock, _, _, Data} = erlfdb_tuple:unpack(V),
+            JobState = job_state(JLock, Seq),
+            {Type, JobId, JobState, decode_data(Data)}
+        end, Result)
+    end).
+
+
+clear_jobs() ->
+    fabric2_fdb:transactional(fun(Tx) ->
+        #{jobs_path := Jobs} = init_jtx(Tx),
+        erlfdb:clear_range_startswith(Tx, Jobs)
+    end).
+
+
+% Private helper functions
+
+maybe_enqueue(#{jtx := true} = JTx, Type, JobId, STime, Resubmit, Data) ->
+    #{tx := Tx} = JTx,
+    Key = job_key(JTx, Type, JobId),
+    JV = #jv{
+        seq = null,
+        jlock = null,
+        stime = STime,
+        resubmit = false,
+        data = Data
+    },
+    case Resubmit of
+        true ->
+            set_job_val(Tx, Key, JV#jv{seq = 0}),
```

Comment:
I'll add a note the top of the fdb module. The reason for the different values is to determine the state of the job from the lock and sequence values.

If lock and sequence are both `null` the job is `finished`
if lock is `not null` then job `running`
if lock is `null` and sequence not null (0) then job is `pending`

So 0 is to distinguish between finished and pending states

Reply:
Seems like an explicit state might be easier? Granted I haven't re-paged in all the details.

Reply:
It's an integer so it's just a different type than null and a version stamp  and it kind of matches with the fact that when a job is added to pending set we bump the `?WATCHES_PENDING` counter.  And when the job is running we set a sequence in the `?ACTIVITY` namespace.


(?JOBS, ?DATA, Type, JobId) = (**SequenceOrCounter**, Lock, SchedTime, Resubmit, JobData)
(?JOBS, ?WATCHES_PENDING, Type) = **Counter**
(?JOBS, ?ACTIVITY, Type, **Sequence**) = JobId

Maybe make it a binary like <<"pending">> instead of 0? or you think an extra explicit state field would be better?

Reply:
I went with a macro, `?PENDING_SEQ = 0`